src/sqlfluff/core/templaters/jinja.py

```diff
@@ -509,6 +520,70 @@ def render_func(in_str: str) -> str:
 
         return env, live_context, render_func
 
+    def _generate_violations_for_undefined_variables(
+        self,
+        in_str: str,
+        syntax_tree: jinja2.nodes.Template,
+        undefined_variables: Set[str],
+    ) -> List[SQLBaseError]:
+        """Generates violations for any undefined variables."""
+        violations: List[SQLBaseError] = []
+        if undefined_variables:
+            # Lets go through and find out where they are:
+            for template_err_val in self._crawl_tree(
+                syntax_tree, undefined_variables, in_str
+            ):
+                violations.append(template_err_val)
+        return violations
+
+    @staticmethod
+    def _init_undefined_tracking(
```

Comment:
Broadly this refactor looks fine. Any reason to declare the UndefinedRecorder class in a function?

Reply:
It needs to depend on a local variable, but there might be a tidier way. At the very least I'll add a better comment, but I'll see if I can find a better way.